### ISSUE TITLE
Add option to use external interpreter to start PyLSP

### ISSUE THREAD

**Issue by Greedquest** [spyder-ide/spyder#20227]
I think this may already be possible. 

### Motivation
It's not ideal that when I use pylsp-mypy to get mypy linting in Spyder, mypy is run in the context of the parent python-lsp-server. Spyder has a dependency on python-lsp-server so if I use conda to install spyder to its own environment, I also need to install pylsp-mypy there (which is a limitation of that library), as well as all my package dependencies to get mypy to work properly.

E.g. if I use GitPython for my project, then I need to install GitPython in the same conda environment as spyder. This pollutes my pristine spyder environment which shouldn't have anything to do with GitPython. Other than aesthetics, this pins some dependencies (e.g. my package is python 3.11 so I want pylsp-mypy to run in a 3.11 environment, but Spyder or some of its dependencies prevent me from upgrading)

### Question

So the question is: Can I install python-lsp-server in my package environment instead and get spyder in my spyder environment to connect to the lsp server unique to my package?

This looks promising
![image](https://user-images.githubusercontent.com/25348920/208510222-d1717207-c553-44e4-8154-3cb488599a86.png)

However I'm not sure about the fields and would really appreciate some help:)

**Comment by ccordoba12** [comment]
Hey @Greedquest, thanks for reporting. You said

> So the question is: Can I install python-lsp-server in my package environment instead and get spyder in my spyder environment to connect to the lsp server unique to my package?

Sure, you can. Just install python-lsp-server, pylsp-mypy and your package dependencies as any other conda package in your environment, then open a terminal and run these commands

    conda activate <my-env>
    python -m pylsp --tcp --port 2090

Finally, you need to go the `Advanced` tab in `Preferences > Completion` and select the following options

![imagen](https://user-images.githubusercontent.com/365293/208721717-ec36e699-5057-478e-abd6-b5b5608e03b9.png)

Let us know if it works for you.

**Comment by Greedquest** [comment]
Nice yep that works great!
Question: does Spyder use TCP or stdio pipes by default? Is there a reason I need the external server checkbox since this stops me from using stdio pipes (which I imagine world be a bit faster maybe)

**Comment by ccordoba12** [comment]
> does Spyder use TCP or stdio pipes by default?

TCP

> Is there a reason I need the external server checkbox since this stops me from using stdio pipes (which I imagine world be a bit faster maybe)

It's because you're connecting Spyder to an external pylsp server (the one you started in your environment). In that case, you can't use stdio pipes for communication.

> I only ask because it feels a little more sluggish than the built in server but maybe that's a caching thing or just imagined.

I think mypy is slow when it starts to analyze files, but @Richardk2n could provide more more insight here.

**Comment by Greedquest** [comment]
Thanks so much @ccordoba12 for all the info, no need for @Richardk2n to get involved as you have already answered all my questions and I think it's just mypy caching if TCP is used for inter process communication to the server either way.


Maybe there's a feature request here. Technically I think that LSP compliance requires the tool (Spyder) to manage the lifetime of the server. 

> The current protocol specification defines that the lifecycle of a server is managed by the client (e.g. a tool like VS Code or Emacs). It is up to the client to decide when to start (process-wise) and when to shutdown a server.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#lifeCycleMessages


This way the tool can supply the port and stuff at startup. Now that creates an issue that difference servers written in different languages will have different ways of launching. VS Code gets around this by forcing every server to also publish a small vscode extension written in Node.js that acts as the client side of the LSP (but is relatively simple to write since vscode provides an API to simplify this)

 I much prefer how Spyder does not require you to write any client code, only the server itself, this feels much more in the language agnostic spirit of LSP. Maybe then it should be possible to provide a command line script for Spyder to launch the server in a child process and supply the port number as a dynamic parameter to the script? 

Either that or a python.exe + module name could be provided to allow running a python language server in a different environment, similar to how you can currently specify the terminal python exe.

Just my thoughts.

**Comment by ccordoba12** [comment]
> Maybe then it should be possible to provide a command line script for Spyder to launch the server in a child process and supply the port number as a dynamic parameter to the script?

That's a good suggestion. But what if we add an option in the `Advanced` tab that allows users to use the same interpreter they set in `Preferences > Python interpreter` to launch the pylsp server?

I think that would be simpler and we'd be able to reuse the same logic we already have to launch pylsp with the default interpreter. At the end, this is what you want, right? I mean, having Spyder in one env and your development packages in another one, right?

**Comment by Greedquest** [comment]
@ccordoba12 Yes that's a good approach and covers my use case. I have been using the custom interpreter for the terminal already, to execute my code in the context of a conda environment different to the one spyder is installed into. Extending this to be used for the pylsp server would be perfect.

**Comment by Richardk2n** [comment]
At that point I would like to interject that this is actually possible in pylsp-mypy (If I understood you correctly).
pylsp-mypy picks the mypy on PATH if there is one. Allowing you to use a mypy in a different env than your spyder env. Also, mypy has an option (python-executeable) that allows you to point it to a different env, which you can provide through a config file or the overrides option of pylsp-mypy.

If you have further questions on that feel free to ask over there.

**Comment by Greedquest** [comment]
@Richardk2n  Aaah, spoilt for choice!

 > pylsp-mypy picks the mypy on PATH if there is one

That is good, however would require modifying PATH any time I switch which project I'm working on, or I'd somehow have to get `conda activate` to happen and run mypy in a conda env... seems tricky.

 > mypy has an option (python-executeable) that allows you to point it to a different env

This is promising, I guess it's a bit nasty having `C:/my/filesystem/conda/env/myenv/python.exe` inside the `mypy.ini` for my project - could lead to the package and the conda env names becoming out of sync. Imagine you have 5 packages all pointing to a conda env then you rename the env, you'd break your mypy integration in a bunch of hard to find places.

I also use pre-commit to run mypy in continuous integration and that would have to ignore the python-executable setting...

I think on balance I prefer the approach @ccordoba12 outlines. It has the added benefit of working not just for mypy but any other lsp setup/envs/plugins. And it gives a single place where I can check the interpreter is correct (spyder) rather than spread over each package. But thanks for the input it's definitely good to know and I would probably use that in the meantime.
